src/librustc_trait_selection/traits/wf.rs

```diff
@@ -312,19 +312,18 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
         let item = self.item;
 
         if let Elaborate::All = elaborate {
-            let predicates = obligations.iter().map(|obligation| obligation.predicate).collect();
-            let implied_obligations = traits::elaborate_predicates(tcx, predicates);
-            let implied_obligations = implied_obligations.map(|pred| {
+            let implied_obligations = traits::util::elaborate_obligations(tcx, obligations.clone());
```

Comment:
At a glance this seems wasteful but it's not worse than it was before. Presumably we can avoid this clone altogether, but I haven't  checked.

Reply:
The only reason I didn't strip them outright was because of `retain` calls in `elaborate_predicates`, but if we can change all cases of `Vec<Pred>` argument types into `impl Iterator<Item = Pred>`, we should be _faster_ than before.